lib_json/src/main/x/json/JsonArrayBuilder.x

```diff
@@ -0,0 +1,173 @@
+     */
+    construct (JsonArray? template = Null, Factory factory = () -> json.newArray()) {
+        this.factory = factory;
+        values = new Array();
+        if (template.is(JsonArray)) {
+            values.addAll(template);
+        }
+    }
+
+    /**
+     * The factory to create a new mutable `JsonObject`.
+     */
+    private Factory factory;
+
+    /**
+     * The array of values to be used to create a JSON array.
+     */
+    private Array<Doc> values;
+
+    /**
+     * @return the number of values that have been added to the builder.
+     */
+    Int size.get() = values.size;
+
+    /**
+     * Add a value to the `JsonArray`.
+     *
+     * @param value  the `Doc` value to add
+     *
+     * @return this `JsonArrayBuilder`
+     */
+    JsonArrayBuilder add(Doc value) {
+        values.add(value);
+        return this;
+    }
+
+    /**
+     * Add the JSON value created by a `JsonBuilder` to the `JsonArray`.
+     *
+     * @param builder  the `JsonBuilder` that will build the `Doc` value to add
+     *
+     * @return this `JsonBuilder`
+     */
+    JsonArrayBuilder add(JsonBuilder builder) = add(builder.build());
+
+    /**
+     * Add all the values to the `JsonArray`.
+     *
+     * @param docs  an array of JSON `Doc` to add
+     *
+     * @return this `JsonBuilder`
+     */
+    JsonArrayBuilder addAll(Doc[] docs) {
+        values.addAll(docs);
+        return this;
+    }
+
+    /**
+     * Set the value at the specified index in the array.
+     *
+     * @param index  the index to add the value at
+     * @param value  the value to set at the specified index
+     */
+    JsonArrayBuilder set(Int index, Doc value) {
+        values.replace(index, value);
+        return this;
+    }
+
+    /**
+     * Set the value produced by a `JsonBuilder` at the specified index in the array.
+     *
+     * @param builder  the `JsonBuilder` that will build the `Doc` value to set
+     * @param value    the value to set at the specified index
+     */
+    JsonArrayBuilder set(Int index, JsonBuilder builder) = set(index, builder.build());
+
+    /**
+     * Build an immutable `JsonArray` from the values added to this builder.
+     */
+    @Override
+    JsonArray build() {
+        JsonArray array = factory();
+        array.addAll(values);
+        return array.makeImmutable();
+    }
+
+    @Override
+    protected Int id(JsonPointer path) = path.index ?: assert;
+
+    @Override
+    protected Doc get(Int index) = values[index];
+
+    @Override
+    protected void update(Int index, Doc doc) = set(index, doc);
+
+    @Override
+    protected void merge(Int index, Doc value) {
+        if (index < 0) {
+            add(value);
+        } else {
+            Doc existing = values[index];
+            switch (existing.is(_)) {
+            case JsonObject:
+                if (value.is(JsonStruct)) {
+                    JsonObject o = new JsonObjectBuilder(existing).deepMerge(value).build();
```

Comment:
Why not to inline this? Then you could further simplify using a ternary or `switch (value.is(_)`